### PR TITLE
System.Configuration port

### PR DESCRIPTION
**Pull request by JeremyKuhne**
Initial change for System.Configuration. This gets it up to the point of a single smoke test.

I've got the change in 2 commits. The first is going through the original sources with automated tools to bring the code closer to standard. The change also includes stripping of unneeded attributes and fixes the string resources. Only exception is I updated TypeHelper.

The **second commit** is the first functional change that strips out the limited trust related code, the Debug class (this was lower than System.Diagnostics), and a few other minor things described in the change.

Still have a lot to do:

1. Writing a ton of tests
2. Finding a good answer for machine.config
3. ??? Other things I'll probably discover in writing tests

### CODE REVIEW

**Comment by JeremyKuhne** [comment]
The general test plan, btw, is to start by writing tests that target commonly documented scenarios (both on MSDN and other sites). Storing application specific metadata is key initial focus.

**Comment by mmitche** [comment]
@dotnet-bot help

**Comment by mmitche** [comment]
@dotnet-bot test Innerloop CentOS7.1 Debug Build and Test

demo

**Comment by mmitche** [comment]
@dotnet-bot help

**Comment by mmitche** [comment]
@dotnet-bot test Innerloop Ubuntu14.04 Debug Build and Test

File: src/System.Configuration/src/System/Configuration/Internal/InternalConfigHost.cs [hunk 1]
```diff
@@ -0,0 +1,325 @@
...
+
+            // ensure the result is in or under the directory of the original source
+            string dirResult = UrlPath.GetDirectoryOrRootName(result);
+            if (!UrlPath.IsEqualOrSubdirectory(dirStream, dirResult))
+                throw new ArgumentException(string.Format(SR.Config_source_not_under_config_dir, configSource));
+
+            return result;
+        }
+
+        internal static object StaticGetStreamVersion(string streamName)
```
Comment (danmosemsft) [inline src/System.Configuration/src/System/Configuration/Internal/InternalConfigHost.cs]: >object [](start = 24, length = 6)

change object to FileVersion?

Comment (JeremyKuhne) [inline src/System.Configuration/src/System/Configuration/Internal/InternalConfigHost.cs]: Sure- from there it gets passed around as object, but no reason to start here.

**Comment by danmosemsft** [comment]
>                         if (factoryRecord != null)

did you intentionally remove this null check?

---
Refers to: src/System.Configuration/src/System/Configuration/BaseConfigurationRecord.cs:1175 in 694b9fe. [](commit_id = 694b9fe639f791198e7771ae1d5c907f532196c6, deletion_comment = True)

**Comment by danmosemsft** [comment]
>         // - Copy the factory and IsFactoryTrustedWithoutAptca bit into the child record

dead comments

---
Refers to: src/System.Configuration/src/System/Configuration/BaseConfigurationRecord.cs:1717 in 8a70508. [](commit_id = 8a70508d46a021a701d2719c0784d425420afca2, deletion_comment = False)

File: src/System.Configuration/src/System/Configuration/Internal/InternalConfigHost.cs [hunk 2]
```diff
@@ -0,0 +1,325 @@
...
+        {
+            if (string.IsNullOrEmpty(streamName))
+                throw new ConfigurationErrorsException(SR.Config_no_stream_to_write);
+
+            // Create directory if it does not exist.
+            // Ignore errors, allow any failure to come when trying to open the file.
+            string dir = Path.GetDirectoryName(streamName);
+
+            if (!Directory.Exists(dir))
+                Directory.CreateDirectory(dir);
```
Comment (JeremyKuhne) [inline src/System.Configuration/src/System/Configuration/Internal/InternalConfigHost.cs]: Probably just an optimization- it doesn't hurt either way.

File: src/System.Configuration/src/System/Configuration/ClientConfigPaths.cs [hunk 3]
```diff
@@ -0,0 +1,359 @@
...
+                if (combinedPath.Length < MaxPath) returnPath = combinedPath;
+            }
+            catch { }
+
+            return returnPath;
+        }
+
+        // Returns a type and hash suffix based on app domain evidence. The evidence we use, in
+        // priority order, is Strong Name, Url and Exe Path. If one of these is found, we compute a 
+        // SHA1 hash of it and return a suffix based on that. If none is found, we return null.
```
Comment (danmosemsft) [inline src/System.Configuration/src/System/Configuration/ClientConfigPaths.cs]: comment out of date

**Comment by JeremyKuhne** [comment]
> did you intentionally remove this null check?

Yeah, it was never null.

> dead comments

Removed

File: src/System.Configuration/src/System/Configuration/ClientConfigPaths.cs (hunk 3 above)
Comment (JeremyKuhne) [inline src/System.Configuration/src/System/Configuration/ClientConfigPaths.cs]: I've clarified the comment.

File: src/System.Configuration/src/System/Configuration/RuntimeConfigurationRecord.cs [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+                            throw new ConfigurationErrorsException(
+                                string.Format(SR.Config_element_locking_not_supported, sectionName), attribute);
+                        }
+                }
+
+                foreach (XmlNode child in xmlNode.ChildNodes)
+                    if (xmlNode.NodeType == XmlNodeType.Element) CheckForLockAttributes(sectionName, child);
+            }
+
+            internal object CreateSection(bool inputIsTrusted, RuntimeConfigurationRecord configRecord,
```
Comment (danmosemsft) [inline src/System.Configuration/src/System/Configuration/RuntimeConfigurationRecord.cs]: >inputIsTrusted [](start = 47, length = 14)

not used. can you remove it throughout?

File: src/System.Configuration/src/System/Configuration/ClientConfigPaths.cs (hunk 3 above)
Comment (JeremyKuhne) [inline src/System.Configuration/src/System/Configuration/ClientConfigPaths.cs]: To be clear- it does the same thing, but using what "used to come from" app domain evidence. This is the same thing I had to do in IsolatedStorage. I'm following up with another change to move the relevant code to common.

File: src/System.Configuration/src/Microsoft/Win32/SafeCryptContextHandle.cs [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+namespace Microsoft.Win32 {
+    using System;
+    using System.Security.Permissions;
+    using Microsoft.Win32.SafeHandles;
+    using System.Configuration;
+
+    // Safehandle for crypt context handles
+    [System.Security.SuppressUnmanagedCodeSecurityAttribute()]
+    internal sealed class SafeCryptContextHandle : SafeHandleZeroOrMinusOneIsInvalid {
+        [SecurityPermission(SecurityAction.LinkDemand, UnmanagedCode = true)]
```
Comment (danmosemsft) [inline src/System.Configuration/src/Microsoft/Win32/SafeCryptContextHandle.cs]: >SecurityPermission [](start = 9, length = 18)

you can remove all the [SecurityPermission] attirbutes

File: src/System.Configuration/src/Resources/Strings.resx [hunk 6]
```diff
@@ -0,0 +1,736 @@
...
+  </data>
+  <data name="ConfigurationPermissionBadXml" xml:space="preserve">
+    <value>Bad Xml {0}</value>
+  </data>
+  <data name="ConfigurationPermission_Denied" xml:space="preserve">
+    <value>Request for ConfigurationPermission failed while attempting to access configuration section '{0}'. To allow all callers to access the data for this section, set section attribute 'requirePermission' equal 'false' in the configuration file where this section is declared.</value>
+  </data>
+  <data name="Section_from_untrusted_assembly" xml:space="preserve">
+    <value>Configuration section '{0}' cannot be instantiated under a partially trusted security policy (AllowPartiallyTrustedCallersAttribute is not present on the configuration section handler).</value>
+  </data>
```
Comment (danmosemsft) [inline src/System.Configuration/src/Resources/Strings.resx]: some strings are dead now -- you may wish to remove them.

File: src/System.Configuration/src/System/Configuration/PropertySourceInfo.cs [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+        {
+            get
+            {
+                // Ensure we return the same string to the caller as the one on which we issued the demand.
+                string filename = _fileName;
+                try
+                {
+                    new FileIOPermission(FileIOPermissionAccess.PathDiscovery, filename).Demand();
+                }
+                catch (SecurityException)
```
Comment (danmosemsft) [inline src/System.Configuration/src/System/Configuration/PropertySourceInfo.cs]: more CAS - are you trying to evict all of it?

File: src/System.Configuration/src/System/Configuration/RuntimeConfigurationRecord.cs (hunk 4 above)
Comment (JeremyKuhne) [inline src/System.Configuration/src/System/Configuration/RuntimeConfigurationRecord.cs]: I'll leave that for a separate item. I did it once then backed out of it- was a little concerned about potential roundtripping issues. I'll create an issue and follow up once I can write some validating tests.

File: src/System.Configuration/src/System/Configuration/PrivilegedConfigurationManager.cs [hunk 8]
```diff
@@ -0,0 +1,27 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Security.Permissions;
+
+namespace System.Configuration
+{
+
+    [ConfigurationPermission(SecurityAction.Assert, Unrestricted = true)]
```
Comment (danmosemsft) [inline src/System.Configuration/src/System/Configuration/PrivilegedConfigurationManager.cs]: >ConfigurationPermission [](start = 5, length = 23)

CAS

File: src/System.Configuration/src/System/Configuration/RuntimeConfigurationRecord.cs (hunk 4 above)
Comment (danmosemsft) [inline src/System.Configuration/src/System/Configuration/RuntimeConfigurationRecord.cs]: I do see that refactoring all dead security stuff would be pulling on a long piece of string and probably not worth it.

---
In reply to: [91443004](https://github.com/dotnet/corefx/pull/14293#discussion_r91443004) [](ancestors = 91443004)

File: src/System.Configuration/src/Microsoft/Win32/SafeCryptContextHandle.cs (hunk 5 above)
Comment (JeremyKuhne) [inline src/System.Configuration/src/Microsoft/Win32/SafeCryptContextHandle.cs]: Heh- cause I didn't put them in the project. :)

File: src/System.Configuration/src/System/Configuration/ClientConfigurationSystem.cs [hunk 9]
```diff
@@ -0,0 +1,220 @@
...
+                    // If we add System.Net.Configuration we'll need to kick the initialization here
+                    // to prevent deadlocks in the networking classes by loading networking config
+                    // before making any networking requests.
+                    //
+                    // Any requests for sections used in initialization during the call to
+                    // EnsureConfigLoaded() will be served by _machine.config or will return null.
+
+                    //if (_isAppConfigHttp)
+                    //{
+                    //}
```
Comment (danmosemsft) [inline src/System.Configuration/src/System/Configuration/ClientConfigurationSystem.cs]: dead code

File: src/System.Configuration/src/Microsoft/Win32/SafeCryptContextHandle.cs (hunk 5 above)
Comment (JeremyKuhne) [inline src/System.Configuration/src/Microsoft/Win32/SafeCryptContextHandle.cs]: Deleted these files, but fixed the ref file.

File: src/System.Configuration/src/Resources/Strings.resx (hunk 6 above)
Comment (JeremyKuhne) [inline src/System.Configuration/src/Resources/Strings.resx]: I'll make a pass

File: src/System.Configuration/src/System/Configuration/IdentityHelper.cs [hunk 10]
```diff
@@ -0,0 +1,134 @@
...
+                    return GetStrongHashSuitableForObjectName(ms);
+                }
+            }
+        }
+
+        internal static string GetStrongHashSuitableForObjectName(string name)
+        {
+            using (MemoryStream ms = new MemoryStream())
+            {
+                using (BinaryWriter b = new BinaryWriter(ms))
```
Comment (danmosemsft) [inline src/System.Configuration/src/System/Configuration/IdentityHelper.cs]: nit: it's shorter to use one set of {}

            using (MemoryStream ms = new MemoryStream())
            using (BinaryWriter b = new BinaryWriter(ms))

File: src/System.Configuration/src/System/Configuration/PropertySourceInfo.cs (hunk 7 above)
Comment (JeremyKuhne) [inline src/System.Configuration/src/System/Configuration/PropertySourceInfo.cs]: Yep, missed it- scanning again

File: src/System.Configuration/src/System/Configuration/RuntimeConfigurationRecord.cs (hunk 4 above)
Comment (danmosemsft) [inline src/System.Configuration/src/System/Configuration/RuntimeConfigurationRecord.cs]: Cleaning up the code in general is of course a non goal in this effort.

File: src/System.Configuration/src/System/Configuration/ClientConfigurationSystem.cs (hunk 9 above)
Comment (JeremyKuhne) [inline src/System.Configuration/src/System/Configuration/ClientConfigurationSystem.cs]: That I left on purpose- to clarify the comment that goes with it.

**Comment by danmosemsft** [comment]
OK, the diff from V1->V2 LGTM.

**Comment by JeremyKuhne** [comment]
Updated based on feedback and removed files I overlooked.

File: src/System.Configuration/ref/System.Configuration.csproj [hunk 11]
```diff
@@ -0,0 +1,17 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
+<Project ToolsVersion="14.0" xmlns="http://schemas.microsoft.com/developer/msbuild/2003" DefaultTargets="Build">
+  <Import Project="$([MSBuild]::GetDirectoryNameOfFileAbove($(MSBuildThisFileDirectory), dir.props))\dir.props" />
+  <PropertyGroup>
+    <OutputType>Library</OutputType>
+    <NuGetTargetMoniker>.NETStandard,Version=v1.7</NuGetTargetMoniker>
+    <ProjectGuid>{2BFEAC8A-DE97-4E50-B992-802C2AD66524}</ProjectGuid>
+    <PackageTargetFramework>netstandard1.7;uap10.1</PackageTargetFramework>
```
Comment (AlexGhiondea) [inline src/System.Configuration/ref/System.Configuration.csproj]: >uap10.1 [](start = 43, length = 7)

is this intentional?

Comment (JeremyKuhne) [inline src/System.Configuration/ref/System.Configuration.csproj]: No- I think I copied that by mistake.

**Comment by AlexGhiondea** [comment]
:shipit:

**Pull request merged**